tsl/src/compression/compression.c

```diff
@@ -2239,118 +2236,559 @@ compressed_insert_key_columns(Relation relation)
+						 errmsg("could not serialize access due to concurrent update")));
+			}
+		}
+		break;
 		/*
-		 * If there are no unique constraints there is nothing to do here.
+		 * If another transaction is updating the compressed data,
+		 * we have to abort the transaction to keep consistency.
 		 */
-		return;
+		case TM_Updated:
+		{
+			elog(ERROR, "tuple concurrently updated");
+		}
+		break;
+		case TM_Invisible:
+		{
+			elog(ERROR, "attempted to lock invisible tuple");
+		}
+		break;
+		case TM_Ok:
+			break;
+		default:
+		{
+			elog(ERROR, "unexpected tuple operation result: %d", result);
+		}
+		break;
 	}
+}
 
-	if (!ts_guc_enable_dml_decompression)
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("inserting into compressed chunk with unique constraints disabled"),
-				 errhint("Set timescaledb.enable_dml_decompression to TRUE.")));
+static inline TM_Result
+delete_compressed_tuple(RowDecompressor *decompressor, Snapshot snapshot,
+						HeapTuple compressed_tuple)
+{
+	TM_FailureData tmfd;
+	TM_Result result;
+	result = table_tuple_delete(decompressor->in_rel,
+								&compressed_tuple->t_self,
+								decompressor->mycid,
+								snapshot,
+								InvalidSnapshot,
+								true,
+								&tmfd,
+								false);
+	return result;
+}
 
-	Oid comp_relid = ts_chunk_get_relid(cis->compressed_chunk_id, false);
-	Relation in_rel = relation_open(comp_relid, RowExclusiveLock);
+struct decompress_batches_stats
+{
+	int64 batches_decompressed;
+	int64 tuples_decompressed;
+};
 
-	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
-	Bitmapset *key_columns = compressed_insert_key_columns(out_rel);
-	Bitmapset *null_columns = NULL;
+/*
+ * This method will:
+ *  1.Scan the index created with SEGMENT BY columns.
+ *  2.Fetch matching rows and decompress the row
+ *  3.insert decompressed rows to uncompressed chunk
+ *  4.delete this row from compressed chunk
+ *
+ *  Returns whether we decompressed anything.
+ */
+static struct decompress_batches_stats
+decompress_batches_using_index(Relation in_rel, Relation out_rel, Relation index_rel,
+							   Snapshot snapshot, ScanKeyData *index_scankeys,
+							   int num_index_scankeys, ScanKeyData *scankeys, int num_scankeys,
+							   Bitmapset *null_columns, List *is_nulls)
+{
+	HeapTuple compressed_tuple;
+	RowDecompressor decompressor;
+	bool decompressor_initialized = false;
+	bool valid = false;
+	int num_segmentby_filtered_rows = 0;
+	int num_heap_filtered_rows = 0;
 
-	int num_scankeys;
-	ScanKeyData *scankeys = build_scankeys(cis->hypertable_relid,
-										   in_rel->rd_id,
-										   &decompressor,
-										   key_columns,
-										   &null_columns,
-										   slot,
-										   &num_scankeys);
+	struct decompress_batches_stats stats = {
+		.batches_decompressed = 0,
+		.tuples_decompressed = 0,
+	};
 
-	bms_free(key_columns);
+	IndexScanDesc scan = index_beginscan(in_rel, index_rel, snapshot, num_index_scankeys, 0);
```

Comment:
Do you think it's possible to move `index_beginscan` one level above (i.e. at chunk insert state initialization), and do only `index_rescan()` when the scankeys change (new inserted tuple)? Begin/end are relatively heavy operations when you repeat them per-tuple.

Reply:
I think its possible, we should also try to batch inserted tuples but I'm going to leave this for upcoming PR just to get this one rolling.

Will add a TODO about this.